### PR TITLE
Full New Listing + first deposit check

### PR DESCRIPTION
**Pull request by nivcertora**
1. **New LLM Workflow for First Deposit**  
   - Introduces `FirstDepositChain` to analyze proposal code and detect first-deposit calls (`approve` + `supply`) for each newly listed asset.  
   - Moves listing logic from `new_listing.py` into this chain, clarifying code responsibilities.

2. **Cached LLM Implementation**  
   - Adds a reusable `CachedLLM` class for Anthropic-based chains, enabling SQLite-based caching of LLM calls.  
   - `IPFSValidationChain` and `FirstDepositChain` both inherit from `CachedLLM`, reducing duplicated setup code.

3. **Prompt & Model Improvements**  
   - Uses a structured output pattern with custom Jinja templates (`first_deposit_prompt.j2`) for coherent extraction of listing details.  
   - Enforces a typed schema (`ListingArray`) to parse LLM output safely, ensuring consistent JSON structures.

4. **Refactored `NewListingCheck`**  
   - Integrates with `FirstDepositChain` to retrieve listings, checks for supply/approval indicators, and logs asset listing results.  
   - Removes local listing classes, reducing duplication and centralizing logic in the chain code.

Successful output:

image

### CODE REVIEW

File: Quorum/llm/prompts/first_deposit_prompt.j2 [hunk 1]
```diff
@@ -0,0 +1,24 @@
+You are an expert Solidity code analyzer specialized in proposals. You will be provided with a snippet of Solidity source code representing a proposal payload that contains new assets listing logic.
```
Comment (itay-certora) [inline Quorum/llm/prompts/first_deposit_prompt.j2:1]: lets change it to "governance proposals"

File: Quorum/llm/prompts/first_deposit_prompt.j2 [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+1. **Asset**: Extract the underlying asset address or identifier from the listing array or the code. This is the token address or name used in the listing struct, often found in code referencing `listings[i].asset` or similar.
+
+2. **Supply/Approval Indicators**: Search for two sequential calls in the code (not necessarily line-by-line) that indicate the following:
+    - An **approval** of the asset to the pool, typically:
+      - `IERC20(asset).approve(...)` or `IERC20(asset).forceApprove(...)`
+      - With arguments including a **pool address** and a **non-zero** amount.
+    - A **supply** (or deposit) call to the pool, typically:
+      - `pool.supply(asset, [non-zero amount], [onBehalfOf], ...)`
+      - Or `pool.deposit(asset, [non-zero amount], [onBehalfOf], ...)`
+
```
Comment (itay-certora) [inline Quorum/llm/prompts/first_deposit_prompt.j2:14]: Although it doesn't have to be line-by-line, the llm could get confused and mix approve and supply of different assets. We should clarify him to find a pair of approve + supply of the same asset

File: Quorum/llm/prompts/first_deposit_prompt.j2 (hunk 1 above)
Comment (nivcertora) [reply Quorum/llm/prompts/first_deposit_prompt.j2:1]: For now its working in the regression test, so if you encounter proposal when it doesn't work as expected, let me know and we can adopt the prompt accordingly

File: Quorum/llm/prompts/first_deposit_prompt.j2 (hunk 2 above)
Comment (nivcertora) [reply Quorum/llm/prompts/first_deposit_prompt.j2:14]: For now its working in the regression test, so if you encounter a proposal when it doesn't work as expected, let me know and we can adopt the prompt accordingly

File: Quorum/checks/new_listing.py [hunk 3]
```diff
@@ -26,9 +13,23 @@ def new_listing_check(self) -> None:
         no new listings were found.
         """
         functions = self._get_functions_from_source_codes()
-        if function := functions.get("newListings", functions.get("newListingsCustom")):
+        if functions.get("newListings", functions.get("newListingsCustom")):
             pp.pretty_print(f"New listings detected for {self.proposal_address}", pp.Colors.WARNING)
-            self._write_to_file("new_listings.json", data=function)
+            proposal_code = self.source_codes[0].file_content
+            proposal_code_str = '\n'.join(proposal_code)
+            listings = FirstDepositChain().execute(proposal_code_str)
```
Comment (yoav-el-certora) [inline Quorum/checks/new_listing.py:20]: Please specify type

File: Quorum/llm/chains/first_deposit_chain.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            ListingDetails object specifying the complete configuration for that
+            particular asset listing.
+    """
+    listings: list[ListingDetails] = Field(description="The list of assets to be listed.")
+
+class FirstDepositChain(CachedLLM):
+
+    def __init__(self):
+        super().__init__()
+        structured_llm = self.llm.with_structured_output(ListingArray)
```
Comment (liav-certora) [inline Quorum/llm/chains/first_deposit_chain.py:55]: Are you sure you can't pass just a list of Listing? That seems pretty weird.

Comment (liav-certora) [reply Quorum/llm/chains/first_deposit_chain.py:55]: Also let me guess, this created an "llm client" that will automatically use the `description` of each field in the pydantic models in the context of each prompt to the llm to return whatever data type you put inside for every message?

File: Quorum/checks/new_listing.py (hunk 3 above)
Comment (nivcertora) [reply Quorum/checks/new_listing.py:20]: done, although i am not sure  why its needed

File: Quorum/llm/chains/first_deposit_chain.py (hunk 4 above)
Comment (nivcertora) [reply Quorum/llm/chains/first_deposit_chain.py:55]: 1. This is the usage according to the docs, it makes sense that it needs a pydantic obj, https://python.langchain.com/docs/how_to/structured_output/
2. Not sure how dark magic works but it does.

**Review by itay-certora — APPROVED** [review]

File: Quorum/llm/chains/first_deposit_chain.py (hunk 4 above)
Comment (liav-certora) [reply Quorum/llm/chains/first_deposit_chain.py:55]: That's cool and very important to know. From the page you linked:
"Beyond just the structure of the Pydantic class, the name of the Pydantic class, the docstring, and the names and provided descriptions of parameters are very important. Most of the time with_structured_output is using a model's function/tool calling API, and you can effectively think of all of this information as being added to the model prompt."

Comment (nivcertora) [reply Quorum/llm/chains/first_deposit_chain.py:55]: This is the easiest and most reliable way to get structured outputs. with_structured_output() is implemented for [models that provide native APIs for structuring outputs](https://python.langchain.com/docs/integrations/chat/), like tool/function calling or JSON mode, and makes use of these capabilities under the hood.

Comment (nivcertora) [reply Quorum/llm/chains/first_deposit_chain.py:55]: https://python.langchain.com/docs/integrations/chat/

Basically its a mix between prompt modification and Inherit support in the LLM tool/ API call that makes it happen

**Pull request merged**